src/appleseed/renderer/modeling/material/imaterialfactory.cpp

```diff
@@ -59,4 +59,71 @@ void IMaterialFactory::add_common_input_metadata(DictionaryArray& metadata)
             .insert("use", "optional"));
 }
 
+void IMaterialFactory::add_alpha_map_metadata(DictionaryArray& metadata)
+{
+    metadata.push_back(
+        Dictionary()
+            .insert("name", "alpha_map")
+            .insert("label", "Alpha Map")
+            .insert("type", "colormap")
+            .insert("entity_types",
+                Dictionary()
+                    .insert("color", "Colors")
+                    .insert("texture_instance", "Textures"))
+            .insert("use", "optional"));
+}
+
+void IMaterialFactory::add_displacement_metadata(DictionaryArray& metadata)
+{
+    metadata.push_back(
+        Dictionary()
+            .insert("name", "displacement_map")
+            .insert("label", "Displacement Map")
+            .insert("type", "colormap")
+            .insert("entity_types",
+                Dictionary().insert("texture_instance", "Textures"))
+            .insert("use", "optional"));
+
+    metadata.push_back(
+        Dictionary()
+            .insert("name", "displacement_method")
+            .insert("label", "Displacement Method")
+            .insert("type", "enumeration")
+            .insert("items",
+                Dictionary()
+                    .insert("Bump Mapping", "bump")
+                    .insert("Normal Mapping", "normal"))
+            .insert("use", "required")
+            .insert("default", "bump")
+            .insert("on_change", "rebuild_form"));
+
+    metadata.push_back(
+        Dictionary()
+            .insert("name", "bump_amplitude")
+            .insert("label", "Bump Amplitude")
+            .insert("type", "numeric")
+            .insert("min_value", "0.0")
+            .insert("max_value", "1.0")
+            .insert("use", "optional")
+            .insert("default", "1.0")
+            .insert("visible_if",
+                Dictionary()
+                    .insert("displacement_method", "bump")));
+
+    metadata.push_back(
+        Dictionary()
```

Comment:
I wonder if there is a way in C++1x to initialize those kind of "dynamic recursive" parameters. All those `insert()` make me nervous thinking to the allocation calls.

Something similar to:

`Dictionary({"name", "normal_map_up"}{"label", "Normal Map Up Vector"} ... )`

Could be nicer.